### PR TITLE
Simple benchmarks

### PR DESCRIPTION
**Pull request by TotalVerb**
It would be helpful to be able to more precisely measure the impact of particular optimizations, so I've imported some benchmarks from the MIT-licensed [nativejson-benchmark](https://github.com/miloyip/nativejson-benchmark). This is very rough, but it's a starting point.

I also added some compliance tests. We're not 100% compliant, but these tests will help pinpoint possible regressions, and also 100% compliance is something to work towards.

### CODE REVIEW

**Comment by kmsquire** [comment]
Haven't looked through thoroughly, but at a glance, 👍 !

**Comment by TotalVerb** [comment]
Since they were there anyway, I added the roundtrips to the tests.

**Comment by tkelman** [comment]
how about downloading a tarball instead of inflating the repo size so dramatically?

**Comment by TotalVerb** [comment]
@tkelman Sure. Will work on that later today.

AppVeyor on 32-bit is apparently failing because of an overflow. That's not really desirable in any case. I think a reasonable patch in the interim would be to use `Int64` always, since that supports all JSON numbers, and fix the numeric parsing later.

**Comment by TotalVerb** [comment]
On second thought, this problem was discovered by a test but not introduced by this PR, so skipping the test seems more reasonable. It will be fixed if number parsing becomes type-stable.

File: test/json-checker.jl [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Run JSON checker tests, excluding tests fail01 and fail18
+
+const JSON_DATA_DIR = joinpath(Pkg.dir("JSON"), "data")
```
Comment (tkelman) [inline test/json-checker.jl]: don't use Pkg.dir, the package may be installed and used from somewhere else. `dirname(@__FILE__)` is better

**Comment by TotalVerb** [comment]
Thanks for the comments. I've made it download the benchmark files when/if required. I also added a citylots file which is much larger.

**Comment by TotalVerb** [comment]
Last try didn't fix the problem. This try should.

**Comment by TotalVerb** [comment]
Well, that still didn't work. I'll have another look tomorrow.

**Comment by TotalVerb** [comment]
Finally passes Appveyor.

File: test/runtests.jl [hunk 2]
```diff
@@ -19,11 +20,12 @@ validate_c(c) = begin
                 end
 
 validate_e(e) = begin
+                    T = TypeVar(:T, AbstractString)
```
Comment (tkelman) [inline test/runtests.jl]: odd, why's this necessary?

Comment (TotalVerb) [inline test/runtests.jl]: I don't think it is. That's an unrelated change that accidentally made it in. I'll take it out.

**Comment by TotalVerb** [comment]
Looks like all tests are passing now. Anything else to handle?

File: test/json-checker.jl [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+for i in [2:17; 19:33]
+    file = "fail$(lpad(i, 2, 0)).json"
+    filepath = joinpath(JSON_DATA_DIR, "jsonchecker", file)
+    try
+        JSON.parsefile(filepath)
+
+        # should not reach here
+        print_with_color(:red, " [Noncompliant] ")
+        println("Expecting $file to fail, but parsed without error")
```
Comment (tkelman) [inline test/json-checker.jl]: should this cause an error?

Comment (TotalVerb) [inline test/json-checker.jl]: Those tests don't work yet though. I would use @test_broken if that were available, but it hasn't landed yet.

Comment (tkelman) [inline test/json-checker.jl]: ah okay. add a #FIXME comment saying to make it an error in the future?

**Comment by tkelman** [comment]
This LGTM aside from the FIXME thing. Will see if anyone else has any opinions, if not ping again if I haven't merged tomorrow.

**Comment by TotalVerb** [comment]
Added two FIXME comments and some better error testing. Also added two extra tests to make the numbers nice (previously we were skipping 1 and 18).

A lot of these will be fixed by the parser optimization I'm working on, so hopefully we'll be able to remove the excess code soon enough.

File: test/json-checker.jl [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    # FIXME: This is a convoluted way of allowing certain tests to fail. Once
+    # ALLOWED_CHECK_FAILURES is empty, we can eliminate the extra code.
+
+    if i ∈ ALLOWED_CHECK_FAILURES
+        try
+            JSON.parsefile(filepath)
+
+            print_with_color(:red, " [Fail] ")
+            println("Test $file parsed successfully; should throw ErrorException.")
+            println(e)
```
Comment (tkelman) [inline test/json-checker.jl]: what's `e` ?

File: test/json-checker.jl [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        catch e
+            if isa(e, ErrorException)
```
Comment (tkelman) [inline test/json-checker.jl]: Doesn't need to be fixed right now, but I wonder if the exception types thrown for various invalid data or parsing failures should be made more specific.

File: test/json-checker.jl (hunk 4 above)
Comment (TotalVerb) [inline test/json-checker.jl]: Oops. Nice catch.

**Comment by TotalVerb** [comment]
These 32-bit issues are kind of annoying to test. I hope this squashes it?

**Comment by tkelman** [comment]
It looks like. What OS do you use?

**Comment by TotalVerb** [comment]
I'm on Linux, but I have access to a Windows box. Both are 64-bit.

**Comment by tkelman** [comment]
You should be able to download the binaries for win32 and linux32 and run them on 64 bit systems. Packages can get confused if there are binary dependencies, but that's not an issue here.

**Comment by nalimilan** [comment]
> how about downloading a tarball instead of inflating the repo size so dramatically?

Does this really make a large difference in cloning time? The files are very small, and less than 300 lines in total.

**Comment by tkelman** [comment]
An earlier version of this PR was adding over 60000 lines. It's fine now.

**Comment by tkelman** [comment]
Thank you for the contribution!

**Pull request merged**